tests/functions/test_pivot_longer.py

```diff
@@ -76,6 +76,472 @@ def test_df():
     )
 
 
+@pytest.fixture
+def df_multi():
+    return pd.DataFrame(
+        {
+            ("name", "a"): {0: "Wilbur", 1: "Petunia", 2: "Gregory"},
+            ("names", "aa"): {0: 67, 1: 80, 2: 64},
+            ("more_names", "aaa"): {0: 56, 1: 90, 2: 50},
+        }
+    )
+
+
+index_labels = [pd.Index(["region"]), {"2007", "region"}]
+column_labels = [{"region": 2007}, {"2007", "2009"}]
+names_to_labels = [1, {12, "newnames"}]
+
+index_does_not_exist = ["Region", [2007, "region"]]
+column_does_not_exist = ["two thousand and seven", ("2007", 2009)]
+
+index_type_checks = (
+    (frame, index) for frame, index in product([df_checks], index_labels)
+)
+column_type_checks = (
+    (frame, column_name)
+    for frame, column_name in product([df_checks], column_labels)
+)
+names_to_type_checks = (
+    (frame, names_to)
+    for frame, names_to in product([df_checks], names_to_labels)
+)
+
+names_to_sub_type_checks = (
+    (df_checks, (1, "rar")),
+    (df_checks, [{"set"}, 20]),
+)
+
+index_presence_checks = (
+    (frame, index)
+    for frame, index in product([df_checks], index_does_not_exist)
+)
+column_presence_checks = (
+    (frame, column_name)
+    for frame, column_name in product([df_checks], column_does_not_exist)
+)
+
+names_sep_not_required = [
+    (df_checks, "rar", "_"),
+    (df_checks, ["blessed"], ","),
+]
+
+names_sep_type_check = [
+    (df_checks, ["rar", "bar"], 1),
+    (df_checks, ("rar", "ragnar"), ["\\d+"]),
+]
+names_pattern_type_check = [
+    (df_checks, "rar", 1),
+    (df_checks, {1: "rar"}, {"\\d+"}),
+]
+
+
+@pytest.mark.parametrize("df,index", index_type_checks)
+def test_type_index(df, index):
+    "Raise TypeError if wrong type is provided for the `index`."
```

Comment:
I think you intend for this line to be a docstring. The same issue occurs in other places. There's a lot of them, so I won't mark them individually but do see if you can fix them up. (If need be, I can hack out a snippet that isolates the functions with non-docstring docstrings, which could help.)